[VPlan] Add scalar inferencing support for Not and Or insns

patrick-rivos opened the pull request:
Fixes #87394.

patrick-rivos committed:
[VPlan] Add scalar inferencing support for Not and Or insns

Fixes #87394.

patrick-rivos commented:
VPlan so @fhahn
Please let me know if this isn't the right approach and I can dig deeper.
There are likely other missing cases but those can be added as the fuzzer finds them.

fhahn commented on llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll:
```diff
@@ -0,0 +1,61 @@
+for.cond.cleanup.loopexit:                        ; preds = %cond.end
+  ret i32 0
+
+for.body:                                         ; preds = %cond.end, %entry
```
nit: something like `loop.header` would be slightly more descriptive

fhahn commented on llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll:
```diff
@@ -0,0 +1,61 @@
+cond.false:                                       ; preds = %for.body
+  br label %cond.end
+
+cond.end:                                         ; preds = %cond.false, %for.body
```
nit: something like `loop.latch` would be slightly more descriptive

fhahn commented on llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll:
```diff
@@ -0,0 +1,61 @@
+for.body:                                         ; preds = %cond.end, %entry
+  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %cond.end ]
+  %zext.0 = zext i1 false to i64
+  %dead.insn = trunc i64 %zext.0 to i16
```
would it be possible to expose the issue without a dead instruction to make the test overall more robust w.r.t to future changes?

fhahn commented on llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll:
```diff
@@ -0,0 +1,61 @@
+
+for.body:                                         ; preds = %cond.end, %entry
+  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %cond.end ]
+  %zext.0 = zext i1 false to i64
```
To make the test more robust w.r.t. future changes, would is it possible to reproduce the issue with a non-loop invariant/constant `zext` here?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:
```diff
@@ -44,8 +45,9 @@ Type *VPTypeAnalysis::inferScalarTypeForRecipe(const VPInstruction *R) {
     CachedTypes[OtherV] = ResTy;
     return ResTy;
   }
+  case VPInstruction::Not:
```
`Not` should only be used to negate booleans (or vectors of bools); Would probably be better to return `IntegerType::get(Ctx, 1);` like for `FCmp/ICmp` and assert that the inferred type of the operand is also `i1`.

(The reason why `PtrAdd` doesn't return `PointerType::get()` but infers the type of the pointer arg is due to address spaces)

fhahn reviewed (commented):
Thanks for the patch!

patrick-rivos committed:
fixup! [VPlan] Add scalar inferencing support for Not and Or insns

patrick-rivos commented:
Thanks for the review!

I split out the not and added an assert. Thankfully the fuzzer has quite a few finds of this ice so I was able to find another testcase that doesn't have any dead variables. The loop-constant seems to be a bit more challenging to get rid of.
I'll check the remaining testcases again just to make sure I didn't miss one.

patrick-rivos committed:
fixup! [VPlan] Add scalar inferencing support for Not and Or insns

fhahn reviewed (approved):
LGTM, thanks!

patrick-rivos commented:
Thank you! I don't have write permission to llvm - could you land this for me?
